src/frontend/src/utils/condition.rs

```diff
@@ -109,6 +109,76 @@ impl Condition {
         .unwrap()
     }
 
+    /// Split the condition expressions into (N choose 2) + 1 groups: those containing two columns
+    /// from different buckets and equal condition between them, and others.
+    ///
+    /// `input_num_cols` are the number of columns in each of the input buckets. For instance, with
+    /// bucket0: col0, col1, col2 | bucket1: col3, col4 | bucket2: col5
+    /// `input_num_cols` = [3, 2, 1]
+    ///
+    /// Returns hashmap with keys of the form (col1, col2) where col1 < col2 in terms of their col
+    /// index.
+    #[must_use]
+    pub fn split_eq_by_input_col_nums(
+        self,
+        input_col_nums: &[usize],
+    ) -> (HashMap<(usize, usize), Self>, Self) {
+        let mut bitmaps = Vec::with_capacity(input_col_nums.len());
+        let mut cols_seen = 0;
+        for cols in input_col_nums {
+            bitmaps.push(FixedBitSet::from_iter(cols_seen..cols_seen + cols));
+            cols_seen += cols;
+        }
+
+        let mut pairwise_conditions = HashMap::with_capacity(self.conjunctions.len());
+        let mut non_eq_join = vec![];
+
+        for expr in self.conjunctions {
+            let input_bits = expr.collect_input_refs(cols_seen);
+            let mut subset_indices = Vec::with_capacity(input_col_nums.len());
+            for (idx, bitmap) in bitmaps.iter().enumerate() {
+                if !input_bits.is_disjoint(bitmap) {
+                    subset_indices.push(idx);
+                }
+            }
+            if subset_indices.len() != 2 || Self::as_eq_cond(&expr).is_none() {
```

Comment:
I guess this could be improved to simply check whether the indexes of the InputRefs belong to two different buckets rather than using the bitmaps.

Reply:
I changed to just be more powerful, we may want to be able to easily split non-eq condition in the future.